File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/safety_fingerprint_v1.py
from __future__ import annotations

import hashlib
from typing import Any

from .canonical_json import canonical_stable_dumps

RULE4_SAFETY_FINGERPRINT_V1 = "rule4-safety-fingerprint-v1"


def build_rule4_safety_fingerprint_v1_payload(
    *,
    safety_gate_status: str,
    safety_status: str,
    prescription_status: str,
    hold_status: str,
    patient_wide_hold: bool,
    urgent_escalation_required: bool,
    analysis_status: str | None,
    d13_hard_stop_active: bool,
    age_verification: str,
    pediatric_band: str | None,
    bp_crisis: bool,
    reason_codes: list[str],
    limitation_codes: list[str],
) -> dict[str, Any]:
    return {
        "fingerprint_version": RULE4_SAFETY_FINGERPRINT_V1,
        "safety_gate_status": safety_gate_status,
        "safety_status": safety_status,
        "prescription_status": prescription_status,
        "hold_status": hold_status,
        "patient_wide_hold": patient_wide_hold,
        "urgent_escalation_required": urgent_escalation_required,
        "analysis_status": analysis_status,
        "d13_hard_stop_active": d13_hard_stop_active,
        "age_verification": age_verification,
        "pediatric_band": pediatric_band,
        "bp_crisis": bp_crisis,
        "reason_codes": sorted(set(reason_codes)),
        "limitation_codes": sorted(set(limitation_codes)),
    }


def rule4_safety_fingerprint_v1_canonical_string(**kwargs: Any) -> str:
    return canonical_stable_dumps(build_rule4_safety_fingerprint_v1_payload(**kwargs))


def rule4_safety_fingerprint_v1_hash(**kwargs: Any) -> str:
    payload = rule4_safety_fingerprint_v1_canonical_string(**kwargs)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest().upper()

File: packages/clinical-contracts/src/rule4/safetyFingerprintV1.ts
import { createHash } from 'node:crypto';

import { canonicalStableDumps } from './canonicalJson.js';

export const RULE4_SAFETY_FINGERPRINT_V1 = 'rule4-safety-fingerprint-v1' as const;

export type Rule4SafetyFingerprintV1Input = {
  safetyGateStatus: string;
  safetyStatus: string;
  prescriptionStatus: string;
  holdStatus: string;
  patientWideHold: boolean;
  urgentEscalationRequired: boolean;
  analysisStatus: string | null;
  d13HardStopActive: boolean;
  ageVerification: string;
  pediatricBand: string | null;
  bpCrisis: boolean;
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
};

export function buildRule4SafetyFingerprintV1Payload(
  input: Rule4SafetyFingerprintV1Input,
): Record<string, unknown> {
  return {
    fingerprint_version: RULE4_SAFETY_FINGERPRINT_V1,
    safety_gate_status: input.safetyGateStatus,
    safety_status: input.safetyStatus,
    prescription_status: input.prescriptionStatus,
    hold_status: input.holdStatus,
    patient_wide_hold: input.patientWideHold,
    urgent_escalation_required: input.urgentEscalationRequired,
    analysis_status: input.analysisStatus,
    d13_hard_stop_active: input.d13HardStopActive,
    age_verification: input.ageVerification,
    pediatric_band: input.pediatricBand,
    bp_crisis: input.bpCrisis,
    reason_codes: [...new Set(input.reasonCodes)].sort(),
    limitation_codes: [...new Set(input.limitationCodes)].sort(),
  };
}

export function rule4SafetyFingerprintV1CanonicalString(
  input: Rule4SafetyFingerprintV1Input,
): string {
  return canonicalStableDumps(buildRule4SafetyFingerprintV1Payload(input));
}

export function rule4SafetyFingerprintV1Hash(input: Rule4SafetyFingerprintV1Input): string {
  const payload = rule4SafetyFingerprintV1CanonicalString(input);
  return createHash('sha256').update(payload, 'utf8').digest('hex').toUpperCase();
}
